File: packages/pvm-assemblyscript/assembly/host-functions/general/read.ts
import { RESULT_CODE_PANIC } from '../../config'
import { CompleteServiceAccount, getStorageValue } from '../../codec'
import {
  ACCUMULATE_ERROR_NONE,
  HostFunctionResult,
} from '../accumulate/base'
import { HostFunctionContext, HostFunctionParams, ReadParams } from './base'
import { BaseHostFunction } from './base'

/**
 * READ host function (Ω_R)
 *
 * Reads storage data from service accounts
 *
 * Gray Paper Specification:
 * - Function ID: 3 (read)
 * - Gas Cost: 10 (plus cold read costs)
 * - Uses registers[7] to specify service account (or self if 2^64-1)
 * - Uses registers[8:3] to specify key offset, key length, and output offset
 * - Reads storage value by key from service account's storage
 * - Writes result to memory at specified offset
 * - Returns NONE if not found, length if found
 */

export class ReadHostFunction extends BaseHostFunction {
  functionId: u64 = u64(3) // READ function ID
  name: string = 'read'
  gasCost: u64 = u64(10)

  execute(
    context: HostFunctionContext,
    params: HostFunctionParams | null,
  ): HostFunctionResult {
    if (!params) {
      context.registers[7] = ACCUMULATE_ERROR_NONE
      return new HostFunctionResult(255)
    }
    const readParams = params as ReadParams

    // Gray Paper equation 404-407: Determine service account
    // s^* = s when registers[7] = 2^64 - 1 (NONE), otherwise registers[7]
    const requestedServiceId =
      u64(context.registers[7]) === ACCUMULATE_ERROR_NONE
        ? readParams.serviceId
        : u64(context.registers[7])

    // Gray Paper equation 408-412: Select service account
    // a = s when s^* = s, otherwise d[s^*] if s^* in keys(d), otherwise none
    let serviceAccount: CompleteServiceAccount | null = null
    if (requestedServiceId === readParams.serviceId) {
      serviceAccount = readParams.serviceAccount
    } else if (readParams.accounts.has(requestedServiceId)) {
      serviceAccount = readParams.accounts.get(requestedServiceId)
    }

    const keyOffset = u64(context.registers[8])
    const keyLength = u64(context.registers[9])
    const outputOffset = u64(context.registers[10])
    const fromOffset = u64(context.registers[11])
    const length = u64(context.registers[12])

    // Read key from memory
    const readResult_key = context.ram.readOctets(u32(keyOffset), u32(keyLength))
    const key = readResult_key.data
    const faultAddress = readResult_key.faultAddress
    if (key === null || faultAddress !== 0) {
      // Gray Paper line 420: registers_7 remains UNCHANGED on panic
      // Do NOT set context.registers[7] - just return PANIC
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // Gray Paper equation 423: Return NONE if service account not found
    if (!serviceAccount) {
      context.registers[7] = ACCUMULATE_ERROR_NONE
      return new HostFunctionResult(255) // continue execution
    }

    // Gray Paper equation 414-418: Read storage value by key using rawCshKeyvals helper
    // v = a_storage[k] if a != none and k in keys(a_storage), otherwise none
    const value = getStorageValue(serviceAccount, u32(requestedServiceId), key!)
    if (!value) {
      // Gray Paper equation 423: Return NONE if storage key not found
      context.registers[7] = ACCUMULATE_ERROR_NONE
      return new HostFunctionResult(255) // continue execution
    }

    // Gray Paper equation 419-420: Calculate slice parameters
    // f = min(registers[11], len(v))
    // l = min(registers[12], len(v) - f)
    const f = min(i32(fromOffset), value.length)
    const l = min(i32(length), value.length - f)
    const dataToWrite = value.slice(f, f + l)

    // Gray Paper equation 421-425: Write to memory and return result
    // Write v[f:l] to memory at offset o (registers[10])
    const writeResult = context.ram.writeOctets(u32(outputOffset), dataToWrite)
    if (writeResult.hasFault) {
      // Gray Paper equation 422: Return panic if memory not writable
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // Gray Paper equation 424: Return len(v) in registers[7]
    context.registers[7] = u64(value.length)

    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/general/read.rs
//! READ host function (Ω_R). Gray Paper: function ID 3.
//! r7=service selector (NONE=self), r8=key offset, r9=key length, r10=output offset, r11=from, r12=length.
//! 1:1 with AS: if !params set r7=NONE and continue; else read key, resolve account, get storage, write slice, r7=len.

use crate::codec::get_storage_value;
use crate::config::{FUNC_READ, REG_NONE};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// READ (3): no params → r7=NONE, continue. With params read key (r8,r9), lookup storage, write at r10 (slice r11,r12).
pub struct ReadHostFunction;

impl HostFunction for ReadHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_READ
    }
    fn name(&self) -> &'static str {
        "read"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let _log_service_id = context.service_id.unwrap_or(0);

        if context.service_id.is_none() || context.accounts.is_none() {
            context.registers[7] = REG_NONE;
            return HostFunctionResult::continue_execution();
        }

        let params_service_id = context.service_id.unwrap();
        let accounts = context.accounts.as_ref().unwrap();

        // Gray Paper equation 404-407: Determine service account
        // s^* = s when registers[7] = 2^64 - 1 (NONE), otherwise registers[7]
        let requested_service_id = if context.registers[7] == REG_NONE {
            params_service_id
        } else {
            context.registers[7]
        };

        // Gray Paper equation 408-412: Select service account
        // a = s when s^* = s, otherwise d[s^*] if s^* in keys(d), otherwise none
        let service_account = if requested_service_id == params_service_id {
            accounts.get(&params_service_id)
        } else {
            accounts.get(&requested_service_id)
        };

        let key_offset = context.registers[8];
        let key_length = context.registers[9];
        let output_offset = context.registers[10];
        let from_offset = context.registers[11];
        let length = context.registers[12];

        let read_key = context.ram.read_octets(key_offset as u32, key_length as u32);
        if read_key.data.is_none() || read_key.fault_address != 0 {
            crate::host_log_error!(
                "[hostfn] read PANIC: key read fault (serviceId={}, offset={}, len={}, fault_address={})",
                _log_service_id, key_offset, key_length, read_key.fault_address
            );
            return HostFunctionResult::panic();
        }
        let key = read_key.data.unwrap();

        let service_account = match service_account {
            Some(a) => a,
            None => {
                context.registers[7] = REG_NONE;
                return HostFunctionResult::continue_execution();
            }
        };
        let value = get_storage_value(service_account, requested_service_id as u32, &key);
        let value = match value {
            Some(v) => v,
            None => {
                crate::host_log_error!(
                    "[hostfn] read: Storage key not found (serviceId={}, requestedServiceId={}, keyLength={})",
                    _log_service_id, requested_service_id, key.len()
                );
                context.registers[7] = REG_NONE;
                return HostFunctionResult::continue_execution();
            }
        };

        let value_len = value.len() as i32;
        let f = (from_offset as i64).min(value_len as i64).max(0) as usize;
        let remaining = (value_len - f as i32).max(0) as usize;
        let l = (length as i64).min(remaining as i64).max(0) as usize;
        let data_to_write = value[f..f + l].to_vec();

        if !data_to_write.is_empty() {
            let write_result = context.ram.write_octets(output_offset as u32, &data_to_write);
            if write_result.has_fault {
                crate::host_log_error!(
                    "[hostfn] read PANIC: output write fault (serviceId={}, offset={}, len={})",
                    _log_service_id, output_offset, data_to_write.len()
                );
                return HostFunctionResult::panic();
            }
        }

        context.registers[7] = value_len as u64;
        HostFunctionResult::continue_execution()
    }
}
